Subscribe with StatefulSet (#3159)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/pubsub/howto-subscribe-statefulset.md

---
type: docs
title: "How to: Subscribe to messages using StatefulSet"
linkTitle: "How to: Subscribe with StatefulSet"
weight: 2000
description: "Learn how to subscribe with StatefulSet and scale horizontally with consistent consumer IDs"
---

> [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) allows deployment of stateful applications on Kubernetes by keeping a sticky identity for each Pod, in contrast to Deployments - where Pods are ephemeral.

Reviewer: ```suggestion
Unlike Deployments, where Pods are ephemeral, [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) allows deployment of stateful applications on Kubernetes by keeping a sticky identity for each Pod.
```

> See an example of a StatefulSet with Dapr:

Reviewer: ```suggestion
Below is an example of a StatefulSet with Dapr:
```

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: python-subscriber
spec:
  selector:
    matchLabels:
      app: python-subscriber  # has to match .spec.template.metadata.labels
  serviceName: "python-subscriber"
  replicas: 3
  template:
    metadata:
      labels:
        app: python-subscriber # has to match .spec.selector.matchLabels
      annotations:
        dapr.io/enabled: "true"
        dapr.io/app-id: "python-subscriber"
        dapr.io/app-port: "5001"
    spec:
      containers:
      - name: python-subscriber
        image: ghcr.io/dapr/samples/pubsub-python-subscriber:latest
        ports:
        - containerPort: 5001
        imagePullPolicy: Always
```

> When subscribing to a pub/sub topic via Dapr, the application can define the `consumerID`, which is used to determine the subscriber's position in the queue or topic. With StatefulSets, developers can take advantage of this sticky identity of Pods and have the `consumerID` to be unique per Pod - allowing each horizontal scale of the subscriber application. Dapr keeps track of the name of each Pod and it can be used when declaring components by using the `{podName}` marker.

Reviewer: ```suggestion
When subscribing to a pub/sub topic via Dapr, the application can define the `consumerID`, which determines the subscriber's position in the queue or topic. With the StatefulSets sticky identity of Pods, you can have a unique `consumerID` per Pod, allowing each horizontal scale of the subscriber application. Dapr keeps track of the name of each Pod, which can be used when declaring components using the `{podName}` marker.
```

> On scaling the number of subscribers of a given topic, each component in Dapr has its unique settings to determine the behavior. Usually, there are two options for multiple consumers: broadcast or shared. In the broadcast configuration, each message published to the topic will be consumed by all subscribers. While in the shared option, a message is consumed by any subscriber (but not all).

Reviewer: ```suggestion
On scaling the number of subscribers of a given topic, each Dapr component has unique settings that determine the behavior. Usually, there are two options for multiple consumers:

 - Broadcast: each message published to the topic will be consumed by all subscribers.
 - Shared: a message is consumed by any subscriber (but not all).
```

> Kafka isolates each subscriber by `consumerID` with its own position in the topic. When an instance restarts, it reuses the same `consumerID` and continues from its last known position, without skipping messages. The component below is an example on how a Kafka component can be used by multiple Pods:

Reviewer: ```suggestion
Kafka isolates each subscriber by `consumerID` with its own position in the topic. When an instance restarts, it reuses the same `consumerID` and continues from its last known position, without skipping messages. The component below demonstrates how a Kafka component can be used by multiple Pods:
```

---
type: docs

> title: "How to: Subscribe to messages using StatefulSet"

Reviewer 2: ```suggestion
title: "How to: Horizontally scale subscribers with StatefulSets"
```

> linkTitle: "How to: Subscribe with StatefulSet"

Reviewer 2: ```suggestion
linkTitle: "How to: Horizontally scale subscribers with StatefulSets"
```
